janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java

```diff
@@ -490,14 +500,20 @@ class BulkRequestChunker implements Iterator<List<RequestBytes>> {
         //There is no "correct" number of actions to perform in a single bulk request. Experiment with different
         // settings to find the optimal size for your particular workload. Note that Elasticsearch limits the maximum
         // size of a HTTP request to 100mb by default
-        private final PeekingIterator<RequestBytes> requestIterator;
+        private final ListIterator<RequestBytes> requestIterator;
         private final int[] exceptionallyLargeRequests;
+        private RequestBytes peeked;
 
         @VisibleForTesting
         BulkRequestChunker(List<ElasticSearchMutation> requests) throws JsonProcessingException {
             List<RequestBytes> serializedRequests = new ArrayList<>(requests.size());
             List<Integer> requestSizesThatWereTooLarge = new ArrayList<>();
-            for (ElasticSearchMutation request : requests) {
+            ListIterator<ElasticSearchMutation> requestsIter = requests.listIterator();
+            while (requestsIter.hasNext()) {
+                ElasticSearchMutation request = requestsIter.next();
+                //Remove the element from the collection so the collection's reference to it doesn't hold it from being
+                //GC'ed after it has been converted to its serialized form
+                requestsIter.set(null);
```

Comment:
Addressing point 1 in https://github.com/JanusGraph/janusgraph/issues/4684

Reply:
This is mutating the `List` that is passed in, which will then be exposed to the caller since we can't take ownership of the List in Java in this context. However from reviewing the call sites, they create the list within their function scopes, and don't do anything with it after passing it into here so this should be "safe", but mutating a passed in list is sometimes frowned upon in Java, so I wanted to call that out.

Call Hierarchy analysis by IntelliJ finds:
- This constructor is only called [here](https://github.com/JanusGraph/janusgraph/blob/4a576f67ff0e53b699ed078dad964d448bc94b10/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java#L551)
- `bulkRequest` is only called [here](https://github.com/JanusGraph/janusgraph/blob/4a576f67ff0e53b699ed078dad964d448bc94b10/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java#L903), [here](https://github.com/JanusGraph/janusgraph/blob/4a576f67ff0e53b699ed078dad964d448bc94b10/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java#L909), [here](https://github.com/JanusGraph/janusgraph/blob/4a576f67ff0e53b699ed078dad964d448bc94b10/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java#L986), and [here](https://github.com/JanusGraph/janusgraph/blob/4a576f67ff0e53b699ed078dad964d448bc94b10/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java#L988)

In each case the passed in List is created in that callsite and isn't mutated, or even read, at least at the time of this writing.